sound/soc/sof/intel/hda-dsp.c

```diff
@@ -759,3 +763,29 @@ int hda_dsp_set_hw_params_upon_resume(struct snd_sof_dev *sdev)
 #endif
 	return 0;
 }
+
+void sof_hda_dsp_d0i3_work(struct work_struct *work)
+{
+	struct sof_intel_hda_dev *hdev = container_of(work,
+						      struct sof_intel_hda_dev,
+						      d0i3_work.work);
+	struct hdac_bus *bus = &hdev->hbus.core;
+	struct snd_sof_dev *sdev = dev_get_drvdata(bus->dev);
+	struct sof_dsp_power_state target_state;
+	int ret;
+
+	target_state.state = SOF_DSP_D0;
+
+	/* DSP can enter D0I3 iff only D0I3-compatible streams are active */
+	if (snd_sof_dsp_only_d0i3_streams_active(sdev))
+		target_state.substate = SOF_HDA_DSP_D0I3;
+	else
+		target_state.substate = SOF_HDA_DSP_D0I0;
+
+	/* This can fail but error cannot be propagated */
+	ret = hda_dsp_set_power_state(sdev, target_state);
```

Comment:
Do we really need to do this if we cannot enter D0i3, aren't we already in D0i0 then?

Reply:
@lyakh sorry didint get your point. We would request an entry to D0i3 from D0i0 but it can fail as you can tell from hda_dsp_set_power_state()? But how do I know even before attempting that it could fail? The point here is that when the system is in S0, we attemp to put the DSP in D0i3 if possible and if not, just leave it in D0i0.

Reply:
@ranj063 exactly, so, why do you have to call `hda_dsp_set_power_state()` to put the DSP into D0i0 if it's already in that state? Or can it now be in a different state? Why?